### PR TITLE
SILKIT-1571 core: add ring buffer for storing data from byte stream

### PR DESCRIPTION
**Pull request by VJonasHolley**
## Subject
Use a ring buffer for storing data from byte stream. 

## Description
This change is introduced for two reasons. First, the message buffer operations w.r.t. reading (e.g. write position) are encapsulated, which results in an improved code readability. Second, a smaller number of copies is necessary when an entire message is passed to the VAsioConnection.

Note that this change does not affect the performance (latency, throughput).

## Instructions for review / testing
- Error handling for write/read: Currently, a boolean is returned (false, if the operation did not succeed). Hence, the class user is responsible for handling an unsuccessful write/read. Is that fine?
- Some member functions are currently unused, but have been useful for the development (write/read for single values, print function). Should those be deleted?
- Testing: Did I miss important corner cases?

## Developer checklist (address before review)

- [ ] Changelog.md updated
- [ ] Prepared update for depending repositories
- [ ] Documentation updated (public API changes only)
- [ ] API docstrings updated (public API changes only)
- [ ] Rebase &rarr; commit history clean
- [ ] Squash and merge &rarr; proper PR title

### CODE REVIEW

File: SilKit/source/core/vasio/RingBuffer.hpp [hunk 1]
```diff
@@ -0,0 +1,352 @@
+/* Copyright (c) 2022 Vector Informatik GmbH
```
Comment (VJanKraemer) [inline SilKit/source/core/vasio/RingBuffer.hpp:1]: You can replace that with the SPDX Header, see the header in https://github.com/vectorgrp/sil-kit/blob/main/SilKit/source/core/vasio/ConnectPeer.cpp

File: SilKit/source/core/vasio/RingBuffer.hpp [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+    , _wPos{0}
+    , _rPos{0}
+{
+    _buffer.resize(capacity);
+}
+
+template <class T>
+void RingBuffer<T>::AdvanceWPos(size_t numBytes)
+{
+    if (numBytes <= 0)
```
Comment (MariusBgm) [inline SilKit/source/core/vasio/RingBuffer.hpp:85]: numBytes is size_t, which is unsigned

File: SilKit/source/core/vasio/VAsioPeer.hpp [hunk 3]
```diff
@@ -121,9 +122,8 @@ class VAsioPeer
 
     // receiving
     std::atomic<uint32_t> _currentMsgSize{0u};
-    std::vector<uint8_t> _msgBuffer;
-    size_t _wPos{0};
-    MutableBuffer _currentReceivingBuffer;
+    RingBuffer<uint8_t> _msgBuffer;
```
Comment (MariusBgm) [inline SilKit/source/core/vasio/VAsioPeer.hpp:125]: Shouldn't we make this a configurable option? Pros: we could easily compare performance, have users switch back, if we suspect trouble. cons: we'd have both code paths lying around.

File: SilKit/source/core/vasio/RingBuffer.hpp [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+template <class T>
+auto RingBuffer<T>::GetArrayTwo() -> RingBuffer<T>::BufArray
+{
+    return BufArray{_buffer.data(), GetSizeArrayTwo()};
+}
+
+template <class T>
+size_t RingBuffer<T>::Capacity()
+{
+    return _capacity;
```
[the remark this answers is not in the document]
Comment (VJanKraemer) [reply SilKit/source/core/vasio/RingBuffer.hpp:236]: If we try to mimic the implementation of std::containers it should actually return 
`return _buffer.capacity()` since capacity is usually about the allocated size, not the numbers of elements the buffer currently holds.

File: SilKit/source/core/vasio/RingBuffer.hpp [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+bool RingBuffer<T>::Full()
+{
+    return _size == _capacity;
+}
+
+template <class T>
+void RingBuffer<T>::SizeCheck()
+{
+    // size must be non-negative
+    if (_size < 0)
```
Comment (VJanKraemer) [inline SilKit/source/core/vasio/RingBuffer.hpp:261]: This should not happen with a `size_t` value

File: SilKit/source/core/vasio/RingBuffer.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (VDanielEdwards) [reply SilKit/source/core/vasio/RingBuffer.hpp:236]: Yes and no, the capacity of the underlying buffer should always be it's size, since this class uses `resize` only to make the buffer bigger. Also the 'capacity' of the ring-buffer must be allocated already.

[the remark this answers is not in the document]
Comment (VJanKraemer) [reply SilKit/source/core/vasio/RingBuffer.hpp:236]: Ah yes, got it. Then we need to use `size()` here, agreed 👍  Thanks for the explanation.

File: SilKit/source/core/vasio/RingBuffer.hpp [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+
+template <class T>
+bool RingBuffer<T>::Write(const T& elem)
+{
+    if (Full()) // we don't allow overwriting
+    {
+        return false;
+    }
+
+    _buffer.at(_wPos) = elem;
```
[the remark this answers is not in the document]
Comment (VJonasHolley) [reply SilKit/source/core/vasio/RingBuffer.hpp]: This code part has been removed since the member function is unused.

File: SilKit/source/core/vasio/RingBuffer.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (VJonasHolley) [reply SilKit/source/core/vasio/RingBuffer.hpp]: Thanks for the hint! I just realized that _capacity is actually redundant.

File: SilKit/source/core/vasio/VAsioPeer.hpp (hunk 3 above)
Comment (VJonasHolley) [reply SilKit/source/core/vasio/VAsioPeer.hpp:125]: I think it would be nice to have only one code path and rely on the ring buffer tests. Do you think the tests cover all important corner cases?

Comment (MariusBgm) [reply SilKit/source/core/vasio/VAsioPeer.hpp:125]: Right, having a single code path and good test coverage is enough.

File: SilKit/source/core/vasio/RingBuffer.hpp [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+public:
+    // public member fcns
+    size_t Capacity();
+    size_t Size();
+
+    bool Read(std::vector<T>& elem, bool advanceRPos = true);
+
+    // write perspective (arrays of free memory in ring buffer)
+    auto GetArrayOne() -> BufArray; // first physically contiguous array (starting at index _wPos)
```
[the remark this answers is not in the document]
Comment (MariusBgm) [reply SilKit/source/core/vasio/RingBuffer.hpp:36]: Could we hide `GetArrayOne` and `GetArrayTwo` behind a `Write(size_t numBytes) -> BufArray` method? it could be called multiple times and return an invalid `{nullptr, 0}` BufArray when buffer space is unavailable.

**Review by MariusBgm — COMMENTED** [review]
Great! I left some questions and ideas. 
Maybe we could simplify the interface of Ringbuffer a bit, and make its operations more obvious.

File: SilKit/source/core/vasio/VAsioPeer.cpp [hunk 8]
```diff
@@ -210,34 +217,29 @@ void VAsioPeer::DispatchBuffer()
...
         {
-            throw SilKitError{"DispatchBuffer: Received message is too small to contain message size header"};
+            throw SilKitError("Reading data from ring buffer failed.");
         }
-        memcpy(&msgSize, _msgBuffer.data(), sizeof msgSize);
-        //ensure buffer does not contain data from contiguous messages
-        _msgBuffer.resize(msgSize);
-        SerializedMessage message{std::move(_msgBuffer)};
+
+        SerializedMessage message{std::move(currentMsg)};
```
Comment (MariusBgm) [inline SilKit/source/core/vasio/VAsioPeer.cpp:239]: We could maybe save one more copy if we refactor SerializedMessage to work on a BufArray (like a vector view). We could advance Rpos after SerializedMessage consumes the date in the BufArray

File: SilKit/source/core/vasio/RingBuffer.hpp [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+public:
+    // constructors and destructors
+    RingBuffer(size_t capacity);
+
+public:
+    // public member fcns
+    size_t Capacity();
+    size_t Size();
+
+    bool Read(std::vector<T>& elem, bool advanceRPos = true);
```
[the remark this answers is not in the document]
Comment (VJonasHolley) [reply SilKit/source/core/vasio/RingBuffer.hpp]: Done. I introduced a fcn 'Peek', which is called in 'Read'. This change makes the optional argument advanceRPos obsolete.

File: SilKit/source/core/vasio/RingBuffer.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (VJonasHolley) [reply SilKit/source/core/vasio/RingBuffer.hpp:37]: Regarding the ring buffer vs double buffering issue: 
I got inspired by the Boost circular_buffer (https://www.boost.org/doc/libs/1_85_0/doc/html/boost/circular_buffer.html) with the arrayOne/arrayOne notation.
image

[the remark this answers is not in the document]
Comment (VJonasHolley) [reply SilKit/source/core/vasio/RingBuffer.hpp:37]: I like the idea of hiding the arrayOne/arrayTwo objects in a 'Write' method. Unfortunately, writing in the VAsioPeer is not performed directly, rather done implicitely by Asio (in the async_read_some method). 

Therefore, I suggest to hide the details in 'ReadSomeAsync' in a method 'SetUpReceivingBuffersForWrite(std::vector<MutableBuffer> _currentReceivingBuffers& buffers)' - or a similar name.

File: SilKit/source/core/vasio/RingBuffer.hpp [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+        AdvanceRPos(elem.size());
+    }
+
+    return true;
+}
+
+template <class T>
+size_t RingBuffer<T>::GetSizeArrayOne()
+{
+    size_t arrayOneSize = std::min(Capacity() - _wPos, Capacity() - _size);
```
[the remark this answers is not in the document]
Comment (VJonasHolley) [reply SilKit/source/core/vasio/RingBuffer.hpp:128]: The variable naming is quite confusing (write vs. read perspective) and definitely needs to be changed. I'll do that in the comment below.

File: SilKit/source/core/vasio/RingBuffer.hpp [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+#include <iostream>
+#include <algorithm>
+#include <cstring>
+
+#include "silkit/participant/exception.hpp"
+
+namespace SilKit {
+namespace Core {
+
+template <typename T>
```
[the remark this answers is not in the document]
Comment (VJonasHolley) [reply SilKit/source/core/vasio/RingBuffer.hpp]: Done. Also, I put all method definitions in a RingBuffer.cpp file (which did not work previously, because of the template parameter).

File: SilKit/source/core/vasio/RingBuffer.hpp [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+
+    void Reserve(size_t newCapacity);
+
+private:
+    // private member fcns
+    bool Empty();
+    void AdvanceRPos(size_t numBytes);
+    void SizeCheck(); // sanity checks
+
+    // write perspective (size of arrays of free memory in ring buffer)
```
[the remark this answers is not in the document]
Comment (VJonasHolley) [reply SilKit/source/core/vasio/RingBuffer.hpp:44]: Done. I chose your suggestion - it might be a bit long, but it should clarify what happens.

File: SilKit/source/core/vasio/VAsioPeer.cpp (hunk 8 above)
Comment (VJonasHolley) [reply SilKit/source/core/vasio/VAsioPeer.cpp:239]: Do you prefer investigating it now or should we put it on a 'Performance ToDo List'/create a ticket?

File: SilKit/source/core/vasio/RingBuffer.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (MariusBgm) [reply SilKit/source/core/vasio/RingBuffer.hpp]: I think giving it a simple name like `auto GetWritingBuffers() -> vector<BufArray>` would be enough to make it obvious (untested pseudo code follows):
```c++
void VAsioPeer::ReadSomeAsync()
{
    SILKIT_ASSERT(_msgBuffer.Capacity() - _msgBuffer.Size() > 0);

    _currentReceivingBuffers.clear();

    auto bufferSequence = _msgBuffer.GetWritingBuffers();
    _currentReceivingBuffers.resize(bufferSequence.size());
    // convert it into asio data structs, or MutableBuffers
    std::transform(bufferSequence.begin(), bufferSequence.end(), _currentReceivingBuffers.begin(),
                       [](const RingBuffer::BufArray& buffer) -> MutableBuffer {
            return MutableBuffer{buffer.first, buffer.second};
    });
    _socket->AsyncReadSome(MutableBufferSequence{_currentReceivingBuffers.data(), _currentReceivingBuffers.size()});
}
```
it could return one of `{}` , `{ArrayOne}`, or `{ArrayOne, ArrayTwo}`. Would that work?

Also it might make sense to re-use MutableBuffer for use in the RingBuffer return values for ArrayOne and Two.

**Review by MariusBgm — APPROVED** [review]
LGTM! thanks for adding the changes I requested

File: SilKit/source/core/vasio/RingBuffer.cpp [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+{
+    // make sure, we only copy as many bytes as are contained in the buffer
+    if (elem.size() > _size)
+    {
+        return false;
+    }
+
+    // copy data from first contiguous array (of occupied memory)
+    size_t numBytesArrayOne = std::min((Capacity() - _rPos), elem.size());
+    memcpy(elem.data(), _buffer.data() + _rPos, numBytesArrayOne);
```
Comment (VDanielEdwards) [inline SilKit/source/core/vasio/RingBuffer.cpp:46]: Not entirely sure why the `std::` namespace was removed.

Afaik the C++ standard does not guarantee that the 'C' names (i.e., in the global namespace) are actually available, it just doesn't forbid it. See https://stackoverflow.com/a/41791540 for a quote from the standard.

I suppose that there are very few compilers that don't provide the functions in the global namespace though.

File: SilKit/source/core/vasio/RingBuffer.cpp [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+    {
+        return false;
+    }
+
+    AdvanceRPos(elem.size());
+
+    return true;
+}
+
+std::vector<MutableBuffer> RingBuffer::GetWritingBuffers()
```
Comment (VDanielEdwards) [inline SilKit/source/core/vasio/RingBuffer.cpp:69]: To me it seems a bit wasteful to always create a new `vector` in each i/o read (whic allocates once or twice each time this method is called!). I'd prefer to pass the `std::vector<MutableBuffer>` as a `&` parameter to the function (similar to `Peek(...)` and `Read(...)`).

File: SilKit/source/core/vasio/RingBuffer.cpp (hunk 13 above)
Comment (VDanielEdwards) [reply SilKit/source/core/vasio/RingBuffer.cpp:46]: This can also just stay as-is as long as MSVC is happy (since I've seen it choke on various functions from `<cstring>` before).

See the `Important` note in [the msdn article on memcpy](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/memcpy-wmemcpy?view=msvc-170#remarks) for the (non-standard) "deprecation" under some circumstances.

**Review by VDanielEdwards — CHANGES_REQUESTED** [review]
I think the `GetWritingBuffers` method should be changed such that it doesn't always create new `std::vector<...>`s.

Otherwise very nice work! 👍

File: SilKit/source/core/vasio/RingBuffer.cpp (hunk 13 above)
Comment (VJonasHolley) [reply SilKit/source/core/vasio/RingBuffer.cpp:46]: Done. Thanks for the hint and explanations!

File: SilKit/source/core/vasio/RingBuffer.cpp (hunk 14 above)
Comment (VJonasHolley) [reply SilKit/source/core/vasio/RingBuffer.cpp:69]: Done.

**Review by VDanielEdwards — APPROVED** [review]
LGTM! 👍
